draft-irtf-cfrg-vdaf.md

```diff
@@ -3977,13 +3977,48 @@ parameter with the same level has been used with the same input share before.
 The whole preparation phase MUST NOT be run more than once for a given
 combination of input share and level.
 
+Also enforces that the prefixes at each level are suffixes of the previous
+level's prefixes, and that the counts of siblings sum up to at most the parent's
+counts.
+
 ~~~
-def is_valid(agg_param, previous_agg_params):
-    (level, _) = agg_param
-    return all(
-        level != other_level
-        for (other_level, _) in previous_agg_params
-    )
+def get_prefix(input, this_level, last_level):
+    # Helper function to determin the prefix of `input` at `last_level`.
+    return input >> (this_level - last_level)
+
+def is_valid(Poplar1, agg_param, previous_agg_params):
+    # Exit early if this is the first time.
+    if len(previous_agg_params) < 1:
+      return True
+
+    (level, prefixes, prefix_counts) = agg_param
+    (last_level, last_prefixes, last_prefix_counts) = previous_agg_params[-1]
+
+    # Check that prefix_count for prefixes are above threshold.
+    for count in prefix_counts:
+      if count < self.threshold:
+        return False
+
+    # Check that level increased.
+    if level <= last_level
+      return False
+
+    # Save last level's prefixes and counts in a dict.
+    last_prefix_count_dict = {}
+    for (i, prefix) in enumerate(last_prefixes):
+      last_prefix_count_dict[prefix] = last_prefix_counts[i]
+
+    # Check that prefixes are suffixes of the last level's prefixes,
+    # and that the counts of siblings sum up to at most the ancestor's count.
+    for (i, prefix) in enumerate(prefixes):
+      last_prefix = get_prefix(prefix, level, last_level)
+      if last_prefix not in last_prefix_count_dict:
+        return False  # Current prefix not a suffix of last level's prefixes.
+      if last_prefix_count_dict[last_prefix] < prefix_counts[i]:
+        return False  # Count exceeds last level's count.
+      last_prefix_count_dict[last_prefix] -= prefix_counts[i]
```

Comment:
I think we also need to check that value in this dict is 0?

Reply:
Removed the dict.